File: frontend/src/api/materials.ts
import client from './client'
import type { ApiResponse } from './client'

export interface MaterialCandidate {
  id: number
  title: string
  content: string
  source_url: string | null
  platform: string
  suggested_material_type: 'reference_article' | 'fact_material'
  status: 'candidate' | 'approved' | 'rejected' | 'failed'
  confidence: number | null
  classification_reason: string | null
  source_kind: string | null
  topic_id: number | null
  article_id: number | null
  source_hash: string | null
  created_at: string | null
  updated_at: string | null
}

export async function getMaterialCandidates(status?: string): Promise<MaterialCandidate[]> {
  const response = await client.get<ApiResponse<MaterialCandidate[]>>('/materials/candidates', {
    params: status ? { status } : {},
  })
  return response.data.data
}

export async function collectMaterials(data: {
  source?: 'topics' | 'articles' | 'hot_articles' | 'search'
  topic_ids?: number[]
  keyword?: string
}): Promise<{ task_id: string }> {
  const response = await client.post<ApiResponse<{ task_id: string }>>('/materials/collect', data)
  return response.data.data
}

export async function approveMaterialCandidate(
  id: number,
  materialType?: 'reference_article' | 'fact_material',
): Promise<{ benchmark_id: number; deduplicated: boolean }> {
  const response = await client.post<ApiResponse<{ benchmark_id: number; deduplicated: boolean }>>(
    `/materials/candidates/${id}/approve`,
    { material_type: materialType },
  )
  return response.data.data
}

export async function rejectMaterialCandidate(id: number): Promise<{ candidate_id: number }> {
  const response = await client.post<ApiResponse<{ candidate_id: number }>>(`/materials/candidates/${id}/reject`)
  return response.data.data
}

File: backend/routes/materials.py
from flask import Blueprint, request

from database import SessionLocal
from models import MaterialCandidate
from utils import error_response, success_response

materials_bp = Blueprint("materials", __name__)


@materials_bp.route("/materials/candidates")
def list_candidates():
    db = SessionLocal()
    try:
        status = request.args.get("status")
        q = db.query(MaterialCandidate).order_by(MaterialCandidate.created_at.desc())
        if status:
            q = q.filter(MaterialCandidate.status == status)
        return success_response([_serialize(candidate) for candidate in q.limit(300).all()])
    except Exception as exc:
        return error_response(str(exc), 500)
    finally:
        db.close()


@materials_bp.route("/materials/collect", methods=["POST"])
def collect_materials():
    from services.material_collection_service import collect_from_articles, collect_from_search, collect_from_topic_ids
    from services.task_manager import task_manager

    body = request.get_json(silent=True) or {}
    source = body.get("source", "topics")
    task_id = task_manager.create_task("material_collect", meta={"source": source})

    if source == "articles":
        task_manager.run(task_id, collect_from_articles, False)
    elif source == "hot_articles":
        task_manager.run(task_id, collect_from_articles, True)
    elif source == "search":
        task_manager.run(task_id, collect_from_search, body.get("keyword", ""))
    else:
        task_manager.run(task_id, collect_from_topic_ids, body.get("topic_ids") or None)
    return success_response({"task_id": task_id})


@materials_bp.route("/materials/candidates/<int:candidate_id>/approve", methods=["POST"])
def approve(candidate_id):
    from services.material_collection_service import approve_candidate

    body = request.get_json(silent=True) or {}
    try:
        return success_response(approve_candidate(candidate_id, body.get("material_type")))
    except Exception as exc:
        return error_response(str(exc), 400)


@materials_bp.route("/materials/candidates/<int:candidate_id>/reject", methods=["POST"])
def reject(candidate_id):
    from services.material_collection_service import reject_candidate

    try:
        return success_response(reject_candidate(candidate_id))
    except Exception as exc:
        return error_response(str(exc), 400)


def _serialize(candidate: MaterialCandidate) -> dict:
    return {
        "id": candidate.id,
        "title": candidate.title,
        "content": candidate.content,
        "source_url": candidate.source_url,
        "platform": candidate.platform,
        "suggested_material_type": candidate.suggested_material_type,
        "status": candidate.status,
        "confidence": candidate.confidence,
        "classification_reason": candidate.classification_reason,
        "source_kind": candidate.source_kind,
        "topic_id": candidate.topic_id,
        "article_id": candidate.article_id,
        "source_hash": candidate.source_hash,
        "created_at": candidate.created_at.isoformat() if candidate.created_at else None,
        "updated_at": candidate.updated_at.isoformat() if candidate.updated_at else None,
    }
